File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_18_Agregar_Color.m
clear;
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'},'File Selector - Imagen A'); 
ruta=strcat(path,file);
A=imread(ruta);
A= im2double(A);
Agris = rgb2gray(A);

% Matrices para las imagenes
[ax ay] = size(Agris);
R1 = zeros(ax, ay, 3);
R2 = zeros(ax, ay, 3);
R3 = zeros(ax, ay, 3);
R4 = zeros(ax, ay, 3);

% azul
vr_1 = -20/255.0;
vg_1 =  8/255.0;
vb_1 = 60/255.0;

% rojo
vr_2=1.4;
vg_2=0.9;
vb_2=0.9;

% naranja
vr_3=1.4;
vg_3=1.15;
vb_3=1.0;

% verde
vr_4= -10/255.0;
vg_4= 40/255.0;
vb_4= -10/255.0;

for m=1:ax
    for n=1:ay
        % Suma azul
        R1(m,n,1) = (vr_1+A(m,n,1));
        R1(m,n,2) = (vg_1+A(m,n,2));
        R1(m,n,3) = (vb_1+A(m,n,3));

        % Multiplicacion rojo
        R2(m,n,1) = (vr_2*A(m,n,1));
        R2(m,n,2) = (vg_2*A(m,n,2));
        R2(m,n,3) = (vb_2*A(m,n,3));
        
        % Multiplicacion naranja
        R3(m,n,1) = (vr_3*A(m,n,1));
        R3(m,n,2) = (vg_3*A(m,n,2));
        R3(m,n,3) = (vb_3*A(m,n,3));
        
        % Suma verde
        R4(m,n,1) = (vr_4+A(m,n,1));
        R4(m,n,2) = (vg_4+A(m,n,2));
        R4(m,n,3) = (vb_4+A(m,n,3));
    end
end

figure;
subplot(231);imshow(A);title('Imagen Original');
subplot(232);imshow(R1);title('Suma(-20,8,60)');
subplot(233);imshow(R2);title('Multi(1.4,0.9,0.9)');
subplot(235);imshow(R3);title('Multi(1.4,1.15,1)');
subplot(236);imshow(R4);title('Suma(-10,40,-10)');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_18_Agregar_Color.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from skimage.color import rgb2gray
from tkinter import Tk, filedialog

file_path = filedialog.askopenfilename(
    title='File Selector - Imagen A',
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path).astype(float) / 255.0

if A.ndim == 3 and A.shape[-1] == 4:
    A = A[..., :3]
elif A.ndim == 2:
    A = np.dstack([A, A, A])

Agris = rgb2gray(A)
ax, ay = Agris.shape

R1 = np.zeros((ax, ay, 3))
R2 = np.zeros((ax, ay, 3))
R3 = np.zeros((ax, ay, 3))
R4 = np.zeros((ax, ay, 3))

vr_1, vg_1, vb_1 = -20/255.0,  8/255.0, 60/255.0   # azul (suma)
vr_2, vg_2, vb_2 = 1.4,        0.9,      0.9       # rojo (multi)
vr_3, vg_3, vb_3 = 1.4,        1.15,     1.0       # naranja (multi)
vr_4, vg_4, vb_4 = -10/255.0, 40/255.0, -10/255.0  # verde (suma)

R1[..., 0] = vr_1 + A[..., 0]
R1[..., 1] = vg_1 + A[..., 1]
R1[..., 2] = vb_1 + A[..., 2]

R2[..., 0] = vr_2 * A[..., 0]
R2[..., 1] = vg_2 * A[..., 1]
R2[..., 2] = vb_2 * A[..., 2]

R3[..., 0] = vr_3 * A[..., 0]
R3[..., 1] = vg_3 * A[..., 1]
R3[..., 2] = vb_3 * A[..., 2]

R4[..., 0] = vr_4 + A[..., 0]
R4[..., 1] = vg_4 + A[..., 1]
R4[..., 2] = vb_4 + A[..., 2]

plt.figure()
plt.subplot(2,3,1); plt.imshow(A);  plt.title('Imagen Original');          plt.axis('off')
plt.subplot(2,3,2); plt.imshow(R1); plt.title('Suma(-20,8,60)');          plt.axis('off')
plt.subplot(2,3,3); plt.imshow(R2); plt.title('Multi(1.4,0.9,0.9)');      plt.axis('off')
plt.subplot(2,3,5); plt.imshow(R3); plt.title('Multi(1.4,1.15,1)');       plt.axis('off')
plt.subplot(2,3,6); plt.imshow(R4); plt.title('Suma(-10,40,-10)');       plt.axis('off')
plt.show()
